#Pending_issue_number Template version is read from the DevonFolder/conf/version.conf …

DSanchezCG opened the pull request:
**Issue number**: Pending to be created
**Description**: If network is not available the template version is set with the value of the fixed constant OASP_TEMPLATE_LAST_STABLE_VERSION. This fixed value might be different from the one in GitHub.

The new changes read the template version from the local file "DevonServerPath/conf/version.json". That file must have a property called "oasp_template_version". That file must be included in future Devon releases.

Regarding the changes:
1 - Added 2 new constants in "Constants.java"
2 - Added 2 new methods in "Utils.java"
3 - Added a new method called "getTemplateVersion" in "oasp4j.java" to encapsulate all the template version logic. If network is available, it reads the  "version.json" from GitHub. Otherwise, it's read from local file "DevonServerPath/conf/version.json". If not found, empty string is returned.

**Note**: Unless there is a big constraint to keep reading "version.json" from GitHub, probably we should stop using it and always read the local  "version.json" file. It's an easy change.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/devonfw/devcon?pullRequest=138) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/devonfw/devcon?pullRequest=138) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/devonfw/devcon?pullRequest=138) it.</sub>

vapadwal commented on src/main/java/com/devonfw/devcon/common/utils/Constants.java:
```diff
@@ -43,6 +43,16 @@
+   * The full path of the version parameters file
+   */
+  public static final String VERSION_PARAMS_FILE_FULL_PATH = "conf" + File.separator + VERSION_PARAMS_FILE;
+
```
I think version.json file is missing in conf directory.

DSanchezCG replied:
Yes, you are right.  As I mentioned in my comments above, these changes suppose the file "version.json" is stored on the local folder "DevonServerPath/conf". Hopefully, this will happen in the next Devon release.

maybeec commented on src/main/java/com/devonfw/devcon/modules/oasp4j/Oasp4j.java:
```diff
@@ -571,4 +570,27 @@ private Node addNode(Document doc, String packagname) {
+    if (oaspTemplateVersion_op.isPresent()) {
+      oaspTemplateVersion = oaspTemplateVersion_op.get();
+    } else {
+      String configPath = Utils.addTrailingSlash(serverPath) + Constants.VERSION_PARAMS_FILE_FULL_PATH;
```
We should implement it in the way, that the configuration takes precedence over all other sources.
Currently, the main issue was that the behavior differs if you are only vs. you are offline. This should not be the case as this causes issues in semi-offline scenarios like trainings.

DSanchezCG replied:
OK! I will change it as you mention: the template version will only be read from local file "version.json"

vapadwal replied:
@maybeec Just confirmation over here that did you mean removing below line 
Optional<String> oaspTemplateVersion_op = Downloader.getDevconConfigProperty(Constants.OASP_TEMPLATE_VERSION);

So now it is only taking the version.json from DevonServerPath/conf whether the network is offline or online . It no more considers the version.json from github. 

I have tested this scenario it throws below error if I dont have version.json in DevonServerPath/conf with network offline/online
Oasp template version not found in config file.

maybeec commented on src/main/java/com/devonfw/devcon/modules/oasp4j/Oasp4j.java:
```diff
@@ -92,10 +92,10 @@ public Oasp4j() {
+    serverpath = serverpath.isEmpty() ? getContextPathInfo().getCurrentWorkingDirectory().toString() : serverpath;
+
+    String oaspTemplateVersion = getTemplateVersion(serverpath);
+    if (oaspTemplateVersion.isEmpty())
```
Should be not empty.
Seems, that this is not tested! Please add tests.

DSanchezCG replied:
I guess it could be empty in rare cases like file missing or file corruption. 
I will add some tests for the "getTemplateVersion" method  creating a temporary folder and a dummy "version.json" file because the "DevonServerPath" location is unknown (AFAIK).

maybeec replied:
I just meant, looking at the diff, it seems, that you invalidated previous business logic. Beforehand, the if clause was checkinf if the version is >not empty<.

DSanchezCG replied:
OK! I will try to keep previous "if" condition by using an Optional<String> datatype

DSanchezCG commented:
Uploaded a new commit having the following changes:
- The method "getTemplateVersion" moved to Utils class
- Added a new method called "removeEndingDot" in Utils class
- Added unit tests for methods "Utils.removeEndingDot" and "Utils.getTemplateVersion"
- Added the file "test/resources/conf/version.json" to perform the unit tests

As I said in the pull request above, these changes suppose the file "version.json" will be included in the "conf" folder of the next Devon release.

maybeec commented on src/test/resources/conf/version.json:
```diff
@@ -0,0 +1,10 @@
+  "oaspide_file_id": "frs54315",
+  "oasp4js_ang1_id": "frs51721",
+  "oasp4js_ang2_id": "frs51961",
+  "oasp_template_version": "2.6.0"
```
maybe better to rename it already or provide two versions here.
oasp4j_template_version / devon4j_template_version. So what is the way to go here is not interesting for me but it should be future proof as of the renaming.

sjimenez77 replied:
IMHO, as devcon is a devonfw tool we should not mix versions. If devon4j is going to be the version that is going to come with the new distribution we will need again to refactor devcon and only refer to the new `devon4j_template_version`.

maybeec commented:
I would expect the following:

1. operate offline + version in conf set to 2.1.0 -> get 2.1.0 archetype
2. operate online + version in conf set to 2.1.0 -> get 2.1.0 archetype
3. operate online + no version set in conf -> get 3.0.0 archetype (latest from internet)
4. operate offline + no version set in conf -> message to the user to go online or setup the config correctly.

DSanchezCG commented:
OK! Looks we are covering all the cases with your proposal. I am implement them asap

maybeec commented:
Thanks, looks good from my side
